### PR TITLE
[L0] Support for Bindless Image Import

### PR DESCRIPTION
**Pull request by nrspruit**

### CODE REVIEW

**Comment by nrspruit** [comment]
Hello @wenju-he ,

Please let me know if you see any issues with the implementation seen here for enabling this functionality.

**Comment by wenju-he** [comment]
> Hello @wenju-he ,
> 
> Please let me know if you see any issues with the implementation seen here for enabling this functionality.

This PR looks good to me.

On DG2 linux, following 3 subtests in https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/bindless_images/vulkan_interop/sampled_images.cpp are passing after applying this PR, https://github.com/intel/llvm/pull/14210 and a change in NEO.
```
  bool valid = run_test<2, float, 4, sycl::image_channel_type::fp32,
                        sycl::image_channel_order::rgba, class float_2d>(
      {16, 16}, {2, 2}, 0);
  valid &= run_test<2, uint32_t, 1, sycl::image_channel_type::unsigned_int32,
                    sycl::image_channel_order::r, class uint32_2d>({64, 32},
                                                                   {4, 2}, 0);
  valid &= run_test<2, int32_t, 1, sycl::image_channel_type::signed_int32,
                    sycl::image_channel_order::r, class int32_2d>({64, 32},
                                                                  {4, 2}, 0);
```
The test uses the three APIs implemented in this PR.
Other subtests have validation errors which are probably not related this PR.

I haven't tested WIN32 path of this PR as our DG2 windows machine, that I can install custom gpu driver, is not usable at the moment. The custom driver needs following two changes to IGC, otherwise device code compilation fails on windows.
1. SYCL bindless image spirv patch are only available in Khronos llvm-spirv translator, but IGC windows build isn't using Khronos llvm-spirv translator, see https://github.com/intel/intel-graphics-compiler/blob/d293f1026ee590d1901b289136942b974f147894/IGC/CMakeLists.txt#L277
So we need to change above line to `ON`.
2. Due to out-of-sync between sycl bindless image header and IGC, we need to either revert https://github.com/intel/intel-graphics-compiler/commit/ae59de16e3ae591014b52e98c5fa5df18391611b or make following change to https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/bindless_images.hpp:
```
diff --git a/sycl/include/sycl/ext/oneapi/bindless_images.hpp b/sycl/include/sycl/ext/oneapi/bindless_images.hpp
index 696301e5c309..8936771e72e5 100644
--- a/sycl/include/sycl/ext/oneapi/bindless_images.hpp
+++ b/sycl/include/sycl/ext/oneapi/bindless_images.hpp
@@ -827,7 +827,11 @@ using OCLImageArrayTyWrite = typename sycl::detail::opencl_image_type<
   __invoke__ImageRead<DataT>(raw_handle, coords)

 #define FETCH_SAMPLED_IMAGE(DataT, raw_handle, coords)                         \
-  __invoke__ImageRead<DataT>(raw_handle, coords)
+  __invoke__ImageReadLod<DataT>(raw_handle, coords, 0.f)
+
+#define SAMPLE_IMAGE_READ(DataT, raw_handle, coords)                           \
+  __invoke__ImageReadLod<DataT>(raw_handle, coords, 0.f)

 #define FETCH_IMAGE_ARRAY(DataT, raw_handle, coords, arrayLayer, coordsLayer)  \
   __invoke__ImageRead<DataT>(raw_handle, coordsLayer)
@@ -846,6 +850,9 @@ using OCLImageArrayTyWrite = typename sycl::detail::opencl_image_type<
 #define FETCH_SAMPLED_IMAGE(DataT, raw_handle, coords)                         \
   __invoke__SampledImageFetch<DataT>(raw_handle, coords)

+#define SAMPLE_IMAGE_READ(DataT, raw_handle, coords)                           \
+  __invoke__ImageRead<DataT>(raw_handle, coords)
+
 #define FETCH_IMAGE_ARRAY(DataT, raw_handle, coords, arrayLayer, coordsLayer)  \
   __invoke__ImageArrayFetch<DataT>(raw_handle, coords, arrayLayer)

@@ -1030,11 +1037,11 @@ DataT sample_image(const sampled_image_handle &imageHandle [[maybe_unused]],

 #ifdef __SYCL_DEVICE_ONLY__
   if constexpr (detail::is_recognized_standard_type<DataT>()) {
-    return __invoke__ImageRead<DataT>(
+    return SAMPLE_IMAGE_READ(DataT,
         CONVERT_HANDLE_TO_SAMPLED_IMAGE(imageHandle.raw_handle, coordSize),
         coords);
   } else {
-    return sycl::bit_cast<DataT>(__invoke__ImageRead<HintT>(
+    return sycl::bit_cast<DataT>(SAMPLE_IMAGE_READ(HintT,
         CONVERT_HANDLE_TO_SAMPLED_IMAGE(imageHandle.raw_handle, coordSize),
         coords));
   }
```

**Review by isaacault — APPROVED** [review]

File: source/adapters/level_zero/image.cpp [hunk 1]
```diff
@@ -947,42 +947,118 @@ UR_APIEXPORT ur_result_t UR_APICALL urBindlessImagesImportExternalMemoryExp(
...
+  externalMemoryData->urMemoryHandle =
+      reinterpret_cast<ur_mem_handle_t>(*phImageMem);
+  return UR_RESULT_SUCCESS;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL urBindlessImagesReleaseInteropExp(
     ur_context_handle_t hContext, ur_device_handle_t hDevice,
     ur_exp_interop_mem_handle_t hInteropMem) {
   std::ignore = hContext;
   std::ignore = hDevice;
```
[the remark this answers is not in the document]
Comment (nrspruit) [reply source/adapters/level_zero/image.cpp:1039]: good point, I will update that in the next patch.

File: source/adapters/level_zero/image.cpp [hunk 2]
```diff
@@ -947,42 +947,118 @@ UR_APIEXPORT ur_result_t UR_APICALL urBindlessImagesImportExternalMemoryExp(
     ur_exp_external_mem_type_t memHandleType,
     ur_exp_interop_mem_desc_t *pInteropMemDesc,
     ur_exp_interop_mem_handle_t *phInteropMem) {
+
   std::ignore = hContext;
   std::ignore = hDevice;
```
[the remark this answers is not in the document]
Comment (nrspruit) [reply source/adapters/level_zero/image.cpp:952]: I am updating with those checks now. This will be fixed in my next update, thanks!

**Comment by nrspruit** [comment]
The e2e failures here:
Failed Tests (9):
  SYCL :: Matrix/SG32/get_coord_int8_matB.cpp
  SYCL :: Matrix/element_wise_all_ops_1d.cpp
  SYCL :: Matrix/element_wise_all_ops_1d_cont.cpp
  SYCL :: Matrix/element_wise_all_ops_scalar.cpp
  SYCL :: Matrix/get_coord_int8_matB.cpp
  SYCL :: Matrix/joint_matrix_apply_bf16.cpp
  SYCL :: Matrix/joint_matrix_apply_two_matrices.cpp
  SYCL :: Matrix/joint_matrix_bfloat16.cpp
  SYCL :: Matrix/joint_matrix_bfloat16_array.cpp
  
  Are all known errors with the system running the e2e tests. These are false negative errors so please ignore.
  
  Let me know if there are any further changes needed to this PR before we add read to merge. Thank you for your time!

**Review by przemektmalon — APPROVED** [review]

**Review by raiyanla — APPROVED** [review]

**Review by winstonzhang-intel — APPROVED** [review]

**Pull request merged**